### ISSUE TITLE
Installation failed with rc6 on macos

### ISSUE THREAD

**Issue by rgaudin** [kiwix/kiwix-hotspot#245]
In GitLab by @Kelson42 on Sep 9, 2018, 11:07

I can start to build the image and things seem to go properly then I got this kind of thing:
![Screen_Shot_2018-09-09_at_13.01.29](/uploads/af6845f77ff83848a0d4afb0c17378dc/Screen_Shot_2018-09-09_at_13.01.29.png)

By reading the log it seems quite clear that something went wrong, but I entry my user/root password in the dialog. Then I get a proper installation failed:
![Screen_Shot_2018-09-09_at_13.03.27](/uploads/b0f57806a82268fcb3fc073de96ac641/Screen_Shot_2018-09-09_at_13.03.27.png)

Remarks:
* I use an external SD card as tmp directory (to write both cache and image)
* I use the wifi to download, so it is not very fast

Here is the whole log
[kelson-222.txt](/uploads/6d67538d340f0addf1c5fc85575b94fe/kelson-222.txt)

**Comment by rgaudin** [comment]
In GitLab by @rgaudin on Sep 9, 2018, 18:08

changed the description

**Comment by rgaudin** [comment]
In GitLab by @rgaudin on Sep 9, 2018, 18:13

Thanks ;

It looks like a download issue as it is clearly stated:

```
Error downloading http://download.kiwix.org/zim/wikisource/wikisource_fr_all_novid_2018-06.zim to /Volumes/Untitled/cache/package_wikisource_fr_all_novid.fr-2018-06-18.zim
("Connection broken: OSError(65, 'No route to host')", OSError(65, 'No route to host'))
```

It asked for your password to restore the mode of the SD-card… and then fires the popup.

Would you have expected a different behavior ? Like the popup first and then the elevation to fix the device mode?

Anyway unless there's a recurrent issue with downloading, this is not a bug… confusing UI I'd say.

**Comment by rgaudin** [comment]
In GitLab by @Kelson42 on Sep 9, 2018, 18:18

Hmmm, so far I know the network is OK, I have been able to create the issue just afterward on my macbook air without (apparent) reconnection. I suspect an instability somehow of the network and a lack of robustess on the HTTP client side... Unfortunately I can not check with a wired connection, the macbook air does not have a RJ45 port.

**Comment by rgaudin** [comment]
In GitLab by @Kelson42 on Sep 9, 2018, 18:36

Have found an adapter, will try with ethernet, and let you know if it works better. Googling a bit the topic of WiFi network interruption on MacOS... it seems to be a pretty common problem. Maybe we should have something a bit more robust. Try again every minute during N minutes and then finally die if the network does not recover?

**Comment by rgaudin** [comment]
In GitLab by @rgaudin on Sep 9, 2018, 18:39

That's most certainly the case ; although it's slightly more robust than it used to be.

We don't log attempts but most certainly this failed after 3 attempts after (1s, 2s, 4s). We could drastically increase this factor if you think it makes sense.

I also use mostly Ethernet

**Comment by rgaudin** [comment]
In GitLab by @Kelson42 on Sep 9, 2018, 20:36

@rgaudin I would try every few seconds, with a log message until the user clicks on abort.

**Comment by rgaudin** [comment]
In GitLab by @rgaudin on Sep 10, 2018, 08:21

Well I'm not in favor of running an infinite loop for this. It seems against our principles for pibox as well.

I will set the following instead: `30s / 60s / 120s` so it's a total of 210s (3.5mn) for a failure.

**Comment by rgaudin** [comment]
In GitLab by @Kelson42 on Sep 10, 2018, 08:35

OK, but please make it longer, we have nothing to win to make it short. Go up to 30mn. You can use Fibonacci if you want ;)

**Comment by rgaudin** [comment]
In GitLab by @rgaudin on Sep 10, 2018, 08:49

Hum OK.

We are using [urllib3.util.retry.Retry](https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html) which handles this for us using the following formula:

```
{backoff factor} * (2 ^ ({number of total retries} - 1))
```

So with a `backoff_factor=30` and a `nb_retries=6` it will wait up to `31.5mn` for a single download. Should it have not come back by then, it's probably safe to fail.

**Comment by rgaudin** [comment]
In GitLab by @Kelson42 on Sep 10, 2018, 08:52

@rgaudin looks good. thx.

**Comment by rgaudin** [comment]
In GitLab by @rgaudin on Sep 11, 2018, 07:42

closed via commit ac3d0a59f2b4b8fc53abf62ae4d7ce3b82760201
